lib/include/jxl/decode.h

```diff
@@ -875,6 +859,71 @@ JXL_EXPORT JxlDecoderStatus JxlDecoderSetPreferredColorProfile(
 JXL_EXPORT JxlDecoderStatus JxlDecoderSetDesiredIntensityTarget(
     JxlDecoder* dec, float desired_intensity_target);
 
+/**
+ * Sets the desired output color profile of the decoded image either from a
+ * color encoding or an ICC profile. Valid calls of this function have either @c
+ * color_encoding or @c icc_data set to NULL and @c icc_size must be 0 if and
+ * only if @c icc_data is NULL.
+ *
+ * Depending of whether a color management system (CMS) has been set the
+ * behavior is as follows:
+ *
+ * If a color management system (CMS) has been set with @ref JxlDecoderSetCms,
+ * and the CMS suppports output to the desired color encoding or ICC profile,
+ * the it will provide the output in that color encoding or ICC profile. If the
+ * desired color encoding or the ICC is not supported, then an error will be
+ * returned.
+ *
+ * If no CMS has been set with @ref JxlDecoderSetCms, there are two cases:
+ *
+ * (1) Calling this function with a color encoding will convert XYB images to
+ * the desired color encoding. In this case, if the requested color encoding has
+ * a narrower gamut, or the white points differ, then the resulting image can
+ * have significant color distortion. Non-XYB images will not be converted to
+ * the desired color space.
+ *
+ * (2) Calling this function with an ICC profile will result in an error.
+ *
+ * In case @ref JxlDecoderSetOutputColorProfile is used with an ICC profile,
+ * (after a in this case required call to JxlDecocderSetCms), the ICC
+ * profile has to be a valid RGB or grayscale color profile.
+ *
+ * Can only be set after the @ref JXL_DEC_COLOR_ENCODING event occurred and
+ * before any other event occurred, and can affect the result of @ref
+ * JXL_COLOR_PROFILE_TARGET_DATA (but not of @ref
+ * JXL_COLOR_PROFILE_TARGET_ORIGINAL), so should be used after getting @ref
+ * JXL_COLOR_PROFILE_TARGET_ORIGINAL but before getting @ref
+ * JXL_COLOR_PROFILE_TARGET_DATA. The color_encoding must be grayscale if
+ * num_color_channels from the basic info is 1, RGB if num_color_channels from
+ * the basic info is 3.
+ *
+ * This function must not be called after JxlDecoderSetCms.
+ *
+ * @param dec decoder orbject
+ * @param color_encoding the output color encoding
+ * @param icc_data bytes of the icc profile
+ * @param icc_size size of the icc profile in bytes
+ * @return @ref JXL_DEC_SUCCESS if the preference was set successfully, @ref
+ *     JXL_DEC_ERROR otherwise.
+ */
+JXL_EXPORT JxlDecoderStatus JxlDecoderSetOutputColorProfile(
+    JxlDecoder* dec, const JxlColorEncoding* color_encoding,
+    const uint8_t* icc_data, size_t icc_size);
+
+/**
+ * Sets the color management system (CMS) that will be used for color
+ * conversion (if applicable) during decoding. May only be set before starting
+ * decoding and must not be called after @ref JxlDecoderSetOutputColorProfile.
+ *
```

Comment:
In the encode api, we also have `If left unset, the default CMS implementation will be used.`. Can we do that here too? Probably only in case `libjxl` is not used and not in case `libjxl_dec` is used.

Reply:
I didn't include using a default CMS, because of the fear that the behavior will be unexpectedly different using different compile time options.
Strong opinions on this, @veluca93?

Reply:
As long as there's a convenient way for applications to get color managed decoding, I'm happy. In my view, defining a custom CMS is pretty advanced usage of the API and it's not reasonable to expect all applications to do that.

I don't think there will be a significant difference in behavior: in case there is no default cms, `JxlDecoderSetOutputColorProfile` will fail on more input jxl bitstreams than if there is one, but an application in any case needs to deal with the case where it fails, so I don't think it makes a big difference. It would only make a difference if an application is using libjxl and doesn't check the result of `JxlDecoderSetOutputColorProfile` (which, to be fair, is something I do expect some applications to do, since as long as they pass it a valid output colorspace, it is not expected to fail with the default CMS), **and then** they switch to libjxl_dec instead of libjxl.

The way I see it (but please correct me if I'm wrong, because I may be misunderstanding something), there are two main use cases for `libjxl_dec`: 1) Chrome and possibly other browsers/viewers that already have their own cms, so they don't want to introduce a dependency on another cms; 2) games and other decode-only applications that only need to decode their own images, so they don't need cms since they can just avoid jxl files with icc profiles.
In the case of 1), they will set a custom cms so there is no issue. In the case of 2), they will not need a cms so there is also no issue.

Most applications will want to depend on `libjxl` though, because they either need encoding too or because they don't care about binary size enough for it to be worth the hassle of figuring out how to do custom cms (or because they are on a sane platform where dynamic linking with a system libjxl is possible, so binary size is not an issue). For those applications, having a convenient way to get a decoded image in the colorspace they want is important imo.

The main issue here is that it's quite a bit of hassle to define a custom cms. I don't mind that people have to add a single line to call `JxlDecoderSetCms()`, but they need to do much more than that, they have to actually implement that whole interface themselves. That means they basically need to write most of `enc_color_management.cc`.

How about we just add a function `JxlDecoderSetDefaultCms()` that either can return failure, or maybe even better, that doesn't exist at all in `libjxl_dec`? That way we can ensure that there is no different behavior, but still have a simple way for application to get access to the default cms if there is one?

Reply:
Having a default function `JxlDecoderSetDefaultCms()` sounds good to me. Added a todo about that.